src/main/java/ch/wisv/areafiftylan/security/SecurityConfiguration.java

```diff
@@ -63,19 +73,43 @@ protected void configure(HttpSecurity http) throws Exception {
             .and().authorizeRequests()
                 .antMatchers("/mail").hasAuthority("ADMIN")
                 .anyRequest().permitAll();
-        //@formatter:on
 
-        // This is used for the Mollie webhook, so it shouldn't be protected by CSRF
-        http.csrf().ignoringAntMatchers("/orders/status");
+        http.csrf().
+            // This is used for the Mollie webhook, so it shouldn't be protected by CSRF
+            ignoringAntMatchers("/orders/status").
+            // Don't require CSRF on requests with valid Tokens
+            requireCsrfProtectionMatcher(csrfRequestMatcher).
+            // We also ignore this for Token requests
+            ignoringAntMatchers("/token");
+        //@formatter:on
 
         // This is the filter that adds the CSRF Token to the header. CSRF is enabled by default in Spring, this just
         // copies the content to the X-CSRF-TOKEN header field.
         http.addFilterAfter(new CsrfTokenResponseHeaderBindingFilter(), CsrfFilter.class);
+
+        // Add support for Token-base authentication
+        http.addFilterAfter(new TokenAuthenticationFilter(authenticationTokenRepository),
+                UsernamePasswordAuthenticationFilter.class);
     }
 
     @Override
     public void configure(AuthenticationManagerBuilder auth) throws Exception {
         auth.userDetailsService(userDetailsService).passwordEncoder(new BCryptPasswordEncoder());
     }
+
+    // We want to enable CSRF for State Changing methods. We don't want CSRF for requests with valid tokens.
+    private RequestMatcher csrfRequestMatcher = new RequestMatcher() {
+        private Pattern allowedMethods = Pattern.compile("^(GET|HEAD|TRACE|OPTIONS)$");
+
+        @Override
+        public boolean matches(HttpServletRequest request) {
+            String requestHeader = request.getHeader("X-Auth-Token");
+            boolean methodAllowed = !(allowedMethods.matcher(request.getMethod()).matches());
+            boolean validAuthToken =
+                    !authenticationTokenRepository.findByToken(requestHeader).filter(Token::isValid).isPresent();
```

Comment:
The `TokenAuthenticationFilter` will also check this later in the chain so this is a little redundant.